### DISCUSSION TITLE
Plugin types aren't being recognized in a Next.js project

### DISCUSSION THREAD

**Discussion by corysimmons** [katspaugh/wavesurfer.js#3531] [Q&A] [answered]
![image](https://github.com/katspaugh/wavesurfer.js/assets/224636/f9d1def5-a427-4073-9113-0796c49bc0ab)

I can see they exist in `node_modules`

![image](https://github.com/katspaugh/wavesurfer.js/assets/224636/2416c8cd-fb1b-463a-9fb4-cac57b4d3acb)

I've also tried installing `@types/wavesurfer.js` and declaring module declarations, but I must be messing something up because it keeps failing to find the types.

They are ultimately `.ts` files so I assume my tsconfig would be able to find them:

![image](https://github.com/katspaugh/wavesurfer.js/assets/224636/8fe07bd9-00b1-4812-9f9c-4c55cec6f8f1)

They seem to just work on the JS example:
![image](https://github.com/katspaugh/wavesurfer.js/assets/224636/552660a6-1c87-4623-948e-be596bfafb78)

So I'm just not sure why they aren't working in a Next.js project.

**Comment by katspaugh** [comment]
You shouldn't install `@types/wavesurfer.js`, it's only for the pre-TypeScript versions (v6 and lower).

Do you have the `wavesurfer.js` package installed as well or only `@wavesurfer/react`?

**Reply by corysimmons** [reply]
> You shouldn't install @types/wavesurfer.js, it's only for the pre-TypeScript versions (v6 and lower).

I suspected.

> Do you have the wavesurfer.js package installed as well or only @wavesurfer/react?

Definitely installed it as well.

**Reply by corysimmons** [reply]
Here's a reduced case https://stackblitz.com/edit/stackblitz-starters-soxkuf

![image](https://github.com/katspaugh/wavesurfer.js/assets/224636/858fb0d1-d870-4c7d-babc-ea0550e0cd48)

**Reply by corysimmons** [reply]
If I add this to my tsconfig it works and I get the types:

```json
"baseUrl": ".",
"paths": {
  "wavesurfer.js/dist/plugins/regions.esm.js": [
    "node_modules/wavesurfer.js/dist/plugins/regions"
  ]
}
```

**Reply by corysimmons** [reply]
I think maybe whatever glob path thing they're using to parse package.json needs all the file extensions on exports.

Doesn't work:
```json
"exports": {
  ".": {
    "import": "./dist/wavesurfer.esm.js",
    "types": "./dist/wavesurfer.d.ts",
    "require": "./dist/wavesurfer.cjs"
  },
  "./dist/plugins/*.js": {
    "import": "./dist/plugins/*.esm.js",
    "types": "./dist/plugins/*.d.ts",
    "require": "./dist/plugins/*.cjs"
  },
  "./plugins/*": {
    "import": "./dist/plugins/*.esm.js",
    "types": "./dist/plugins/*.d.ts",
    "require": "./dist/plugins/*.cjs"
  },
  "./dist/*": {
    "import": "./dist/*"
  }
}
```

Works:
```json
"exports": {
  ".": {
    "import": "./dist/wavesurfer.esm.js",
    "types": "./dist/wavesurfer.d.ts",
    "require": "./dist/wavesurfer.cjs"
  },
  "./dist/plugins/*.js": {
    "import": "./dist/plugins/*.esm.js",
    "types": "./dist/plugins/*.d.ts",
    "require": "./dist/plugins/*.cjs"
  },
  "./plugins/*": {
    "import": "./dist/plugins/*.esm.js",
    "types": "./dist/plugins/*.d.ts",
    "require": "./dist/plugins/*.cjs"
  },
  "./dist/*": {
    "import": "./dist/*"
  },
  "./dist/plugins/*.esm.js": {
    "import": "./dist/plugins/*.esm.js",
    "types": "./dist/plugins/*.d.ts",
    "require": "./dist/plugins/*.cjs"
  }
}
```

Notice how you were doing `./dist/plugins/*.js` and adding the `.esm.js` fixed it `./dist/plugins/*.esm.js`

**Reply by katspaugh** [reply]
True, the way those exports are currently specified the import should probably be “regions.js” instead of “regions.esm.js”.

**Reply by corysimmons** [reply, marked as answer]
https://github.com/katspaugh/wavesurfer.js/pull/3533
